core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java

```diff
@@ -0,0 +1,1059 @@
+          cancellationGizmo
+      );
+
+      getPool().execute(finalMergeAction);
+    }
+
+    /**
+     * Computes maximum number of layer 1 parallel merging tasks given available processors and an estimate of current
+     * {@link ForkJoinPool} utilization. A return value of 1 or less indicates that a serial merge will be done on
+     * the pool instead.
+     */
+    int computeNumTasks()
+    {
+      // max is minimum of either number of processors or user suggested parallelism
+      final int maxParallelism = Math.min(JvmUtils.getRuntimeInfo().getAvailableProcessors(), parallelism);
+      // adjust max to be no more than total pool parallelism less the number of running threads + submitted tasks
+      final int utilizationEstimate = getPool().getRunningThreadCount() + getPool().getQueuedSubmissionCount();
+      // minimum of 'max computed parallelism' and pool parallelism less current 'utilization estimate'
+      final int computedParallelism = Math.min(maxParallelism, getPool().getParallelism() - utilizationEstimate);
+      // compute total number of layer 1 'parallel' tasks, the final merge task will take the remaining slot
+      final int computedOptimalParallelism = Math.min(
+          (int) Math.floor((double) sequences.size() / 2.0),
+          computedParallelism - 1
+      );
+      return computedOptimalParallelism;
+    }
+  }
+
+
+  /**
+   * This {@link RecursiveAction} is the work-horse of the {@link ParallelMergeCombiningSequence}, it merge-combines
+   * a set of {@link BatchedResultsCursor} and produces output to a {@link BlockingQueue} with the help of a
+   * {@link QueuePusher}. This is essentially a composite of logic taken from {@link MergeSequence} and
+   * {@link org.apache.druid.common.guava.CombiningSequence}, where the {@link Ordering} is used to both set the sort
+   * order for a {@link PriorityQueue}, and as a comparison to determine if 'same' ordered results need to be combined
+   * with a supplied {@link BinaryOperator} combining function.
+   *
+   * This task takes a 'yieldAfter' parameter which controls how many input result rows will be processed before this
+   * task completes and executes a new task to continue where it left off. This value is initially set by the
+   * {@link MergeCombinePartitioningAction} to a default value, but after that this process is timed to try and compute
+   * an 'optimal' number of rows to yield to achieve a task runtime of ~10ms, on the assumption that the time to process
+   * n results will be approximately the same.
+   */
+  private static class MergeCombineAction<T> extends RecursiveAction
+  {
+    private final PriorityQueue<BatchedResultsCursor<T>> pQueue;
+    private final Ordering<T> orderingFn;
+    private final BinaryOperator<T> combineFn;
+    private final QueuePusher<ResultBatch<T>> outputQueue;
+    private final T initialValue;
+    private final int yieldAfter;
+    private final int batchSize;
+    private final int depth;
```

Comment:
Suggest adding a comment about how `depth` is the recursion count and is used for cumulative moving average smoothing as well